src/ethereum/block-sync.ts

```diff
@@ -71,7 +57,18 @@ export class BlockSync {
     return (await this.reader.getBlockNumber()) - this.config.FinalityBufferBlocks;
   }
 
+  // Note about tracking changes in contract registry:
+  // If contract address was updated in contract registry in the middle of block 1000,
+  // we read blocks 1-1000 from old address and blocks 1001+ from the new address.
+  // This simplification is ok because contracts will be locked from emitting events during transition.
+
   async processEventsInBlock(blockNumber: number, latestAllowedBlock: number) {
+    // update all contract addresses according to state to track changes in contract registry
+    for (const eventName of eventNames) {
+      const address = this.state.getCurrentSnapshot().CurrentContractAddress[contractByEventName(eventName)];
+      if (address) this.eventFetchers[eventName].setContractAddress(address);
+    }
```

Comment:
That was my intention originally but it kept getting complicated. I talked to Oded and he wanted me to cut cleanly in the middle of a block, meaning remove the extra events if found so we have a clean switch in the middle. Since the derived classes had to be aware of this logic and clear their caches/lookaheads, it got a little too complicated so I told Oded to simplify the whole thing - so he said we have locking so no actual events will appear anyways - so we don't need to spend so much effort on this. So I dropped the whole angle.

If you think this can be improved without adding complexity, I'll be happy to see a PR. I couldn't imagine a simple way but I didn't try too hard either.